# Bioinformatics

## Bioinformatics defination

Bioinformatics: an interdisciplinary field that develops and applies computer and computational technologies to study biomedical questions.
- As a technology, bioinformatics is a powerful technology to manage, query, and analyze big data in life sciences
- As a methodology, bioinformatics is a top-down, holistic, data-driven, genome-wide, and systems approach that generates new hypotheses, finds new patterns, and discovers new functional elements

## History of Bioinformatics

**The Bio- in Bioinformatics**

**Genotype ----> Phenotype**

DNA/Genome --> RNA --> Proteins --> Molecular Networks --> Cells --> Physiology/Disease

**Data ----> Discovery**

Data management --> Data computation --> Data mining --> Modeling/Simulation

History
- 1945, first electronic computer, ENIAC
- 1948, Shannon information theory
- 1953, **DNA double-helix structure**
- 1953, **first protein sequence determined**
- 1953, **first protein structure solved**
- 1953, first commerical (mass-produced) computers
- 1953, theory of games, grammars, FORTRAN
- 1962-1965, **molecular evolutionary clock**
- 1962-1965, cellular automata, theory of computation
- 1969, ARPANET created
- 1969, Email, Ethernet and TCP described
- 1977, **Sanger DNA sequencing**
- 1980-1990, personal computer, DNS launched
- 1990, **human genome project started**
- 1995, **H. influenzae genome sequenced**
- **E.coli, yeast, worm, fly genomes**
- 2001/2004, **human genomes sequenced**
- 2005, **Next-generation sequencing**
- 1951-1952, **1st computer program to determine protein structure**
- 1962, COMPROTEIN, M.O. Dayhoff
- 1965, protein atlas
- 1966, evolution of proteins
- 1967, phylogenetic tree
- 1970, first apperances of the word 'Bioinformatics': Hesper B., Hogeweg P. (1970). "Bioinformatica: een wekconcept", Kameleon, 1(6): 28-29. "The study of informatic processes in biotic systems"
- 1970, sequence alignment by DP
- 1971, **Protein Data Bank (PDB)**
- 1974, **Chou-Fasman 2nd structure prediction**
- 1975, **Levitt-Warshel computer simulation of protein folding**
- 1978, PAM substitution matrix, BLOSUM (1992)
- 1981, Smith-Waterman local alignment
- 1982, Genbank
- 1983, Protein Information Resource (PIR)
- 1990, 1997, BLAST, Gapped BLAST, PSI BLAST
- 1994, **CASP structure prediction assessment**
- Microarray gene expression analysis
- Gene prediction from genome
- Genome alignment
- 2002, BLAT
- 2007, SRA
- NGS data analysis

Examples of contributions to computer sciences
- Neural networks
- Genetic programming
- WAN

## Bioinformatics in Mainland China

Rapid growth since 1990s. Driving force: 1. Internet 2. Genomics 3. increased research funding (Ministry of Science and Technology (MOST), Natural Science Foundation of China (NSFC)) 4. critical mass of researchers & students.

Reading materials:

Required:

Bioinformatics. 2003 Nov 22;19(17):2176-90. Early bioinformatics: the birth of a discipline--a personal view. Ouzounis CA, Valencia A. PMID: 14630646
PLoS Comput Biol. 2011 Mar;7(3):e1002021. doi: 10.1371/journal.pcbi.1002021. Epub 2011 Mar 31. The roots of bioinformatics in theoretical biology. Hogeweg P. PMID: 21483479
PLoS Comput Biol. 2008 Apr 25;4(4):e1000020. doi: 10.1371/journal.pcbi.1000020. Bioinformatics in China: a personal perspective. Wei L, Yu J. PMID: 18437216
Optional:
Nature. 1991 Jan; 349(10):99 Towards a paradigm shift in biology. Gillbert W.

## Sequance Alignment

Readings

Required

J Mol Biol. 1981 Mar 25;147(1):195-7. Identification of common molecular subsequences. Smith TF, Waterman MS. PMID: 7265238(Note: you might need a subscription to the Journal of Molecular Biology magazine in order to view the full-text of this paper. However you might also find other websites holding the full-text for this paper, because it is so famous.)
J Mol Biol. 1970 Mar;48(3):443-53. A general method applicable to the search for similarities in the amino acid sequence of two proteins. Needleman SB, Wunsch CD. PMID: 5420325 (Note: you might need a subscription to the Journal of Molecular Biology magazine in order to view the full-text of this paper. However you might also find other websites holding the full-text for this paper, because it is so famous.)
Optional
Proc Natl Acad Sci U S A. 1992 Nov 15; 89(22): 10915–9. Amino acid substitution matrices from protein blocks. S Henikoff and J G Henikoff. PMID: 1438297
Nature. 1953 Apr 25;171(4356):737-8.Molecular structure of nucleic acids; a structure for deoxyribose nucleic acid. WATSON JD, CRICK FH. PMID: 13054692 (Note: this paper is so old, which seems to be the reason why PubMed does not provide a full-text link for this paper. Nevertheless, this paper is so famous, and you can find it easily on the Internet (and even on Wikipedia! Try it.) . Here is the link to the full-text of this paper on the website of Nature Publishing Group: http://www.nature.com/nature/dna50/watsoncrick.pdf)
Horizons in Biochemistry. 1962 pp. 189-225. Molecular disease, evolution, and genetic heterogeneity. Zuckerkandl E., Pauling L.B. (Note: this paper is not archived in PubMed. I (Yang Ding the TA) cannot find the full-text of paper online as I'm writing this page, and since there are copyright issues, I recommend you to refer to some library to find the full-text of this paper. The format of reference here has not been confirmed to be correct yet, and we will check it later.)
Acta Chem. Scand. 1963;17:S9-S16. Pauling L., Zuckerkandl E. Chemical paleogenetics: molecular "restoration studies" of extinct forms of life (Note: this paper is not archived in PubMed. I (Yang Ding the TA) cannot find the full-text of paper online as I'm writing this page, and since there are copyright issues, I recommend you to refer to some library to find the full-text of this paper.)
Evolving Genes and Proteins. 1965 pp. 97-166. Zuckerkandl E., Pauling L. Evolutionary divergence and convergence in proteins (Note: this paper is not archived in PubMed. You can, however, access the full-text of this paper from here: http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.550.4751&rep=rep1&type=pdf
Kameleon. 1970 1(6);28-29. Hesper B., Hogeweg P. Bioinformatica: een werkconcept.
Simulation. 1978 Sep 1;31(3):90. Hogeweg, P. Simulating the growth of cellular forms. (Note: this paper is not archived in PubMed. I (Yang Ding the TA) cannot find the full-text of paper online as I'm writing this page, and since there are copyright issues, I recommend you to refer to some library to find the full-text of this paper.)
Acta Cryst. 1952;5:109-116. Bennett J.M., Kendrew J.C. The computation of fourier syntheses with a digital electronic calculating machine.(Note: this paper is not archived in PubMed. I (Yang Ding the TA) cannot find the full-text of paper online as I'm writing this page, and since there are copyright issues, I recommend you to refer to some library to find the full-text of this paper.)
Proceedings-Fall Joint Computer Conference. 1962. Dayhoff M., Ledley R. COMPROTEIN: a computer program to aid primary protein structure determination. (Note: this paper is not archived in PubMed. I (Yang Ding the TA) cannot find the full-text of paper online as I'm writing this page, and since there are copyright issues, I recommend you to refer to some library to find the full-text of this paper.)
Science. 1966 Apr 15;152(3720):363-6. Eck RV, Dayhoff MO. Evolution of the structure of ferredoxin based on living relics of primitive amino Acid sequences. PMID: 17775169 (Note: you might need a subscription to the Science magazine in order to view the full-text of this paper.)
Science. 1967 Jan 20;155(3760):279-84. Fitch WM, Margoliash E.Construction of phylogenetic trees. PMID: 5334057 (Note: you might need a subscription to the Journal of Molecular Biology magazine in order to view the full-text of this paper.)
J Mol Biol. 1970 Mar;48(3):443-53. A general method applicable to the search for similarities in the amino acid sequence of two proteins. Needleman SB, Wunsch CD. PMID: 5420325 (Note: you might need a subscription to the Journal of Molecular Biology magazine in order to view the full-text of this paper.)
Atlas of protein sequence and structure. 1978. Chapter 22: A model of evolutionary change in proteins. Dayhoff M. O., Schwartz R. M. (Note: this paper is not archived in PubMed. I (Yang Ding the TA) cannot find the full-text of paper online as I'm writing this page, and since there are copyright issues, I recommend you to refer to some library to find the full-text of this paper.)
J Mol Biol. 1981 Mar 25;147(1):195-7. Identification of common molecular subsequences. Smith TF, Waterman MS. PMID: 7265238 (Note: you might need a subscription to the Journal of Molecular Biology magazine in order to view the full-text of this paper. However you might also find other websites holding the full-text for this paper, because it is so famous.)
J Mol Biol. 1990 Oct 5;215(3):403-10. Basic local alignment search tool. Altschul SF, Gish W, Miller W, Myers EW, Lipman DJ. PMID: 2231712 (Note: you might need a subscription to the Journal of Molecular Biology magazine in order to view the full-text of this paper. However you might also find other websites holding the full-text for this paper, because it is so famous.)
Nucleic Acids Res. 1997 Sep 1;25(17):3389-402. Gapped BLAST and PSI-BLAST: a new generation of protein database search programs. Altschul SF, Madden TL, Schäffer AA, Zhang J, Zhang Z, Miller W, Lipman DJ. PMID: 9254694
Bioinformatics. 2003 Nov 22;19(17):2176-90. Early bioinformatics: the birth of a discipline--a personal view. Ouzounis CA, Valencia A. PMID: 14630646

## Sequence Database Search

Readings

Required

J Mol Biol. 1990 Oct 5;215(3):403-10. Basic local alignment search tool. Altschul SF, Gish W, Miller W, Myers EW, Lipman DJ. 

(Note: currently there is no full-text link for this article in PubMed. However you might also find other websites holding the full-text for this paper, because it is so famous.)

http://www.fi.muni.cz/~lexa/blast.html

Nucleic Acids Res. 1997 Sep 1;25(17):3389-402. Gapped BLAST and PSI-BLAST: a new generation of protein database search programs. Altschul SF, Madden TL, Schäffer AA, Zhang J, Zhang Z, Miller W, Lipman, D.J. PMID: 9254694

Genome Biol. 2001;2(10):REVIEWS2002. Having a BLAST with bioinformatics (and avoiding BLASTphemy). Pertsemlidis A, Fondon JW 3rd. PMID: 11597340

Nucleic Acids Res. 2006 Jul 1;34(Web Server issue):W6-9. BLAST: improvements for better sequence analysis. Ye J, McGinnis S, Madden TL. PMID: 16845079

Optional

Nat Genet. 1993 Mar;3(3):266-72. Identification of protein coding regions by database similarity search. Gish W, States DJ. PMID: 8485583

Note: this paper is supplementary and thus OPTIONAL. You might need a subscription to the Nature Genetics magazine in order to view the full-text of this paper.

Proc Natl Acad Sci U S A. 1990 Mar;87(6):2264-8. Methods for assessing the statistical significance of molecular sequence features by using general scoring schemes. Karlin S, Altschul SF. PMID: 2315319

Note: this paper is supplementary and thus OPTIONAL.

Proc Natl Acad Sci U S A. 1993 Jun 15;90(12):5873-7. Applications and statistics for multiple high-scoring segments in molecular sequences. Karlin S, Altschul SF. PMID: 8390686

Note: this paper is supplementary and thus OPTIONAL.

Methods Enzymol. 1996;266:554-71. Analysis of compositionally biased regions in sequence databases. Wootton JC, Federhen S. PMID: 8743706

Note: this paper is supplementary and thus OPTIONAL. You might need a subscription to the Methods in Enzymology magazine in order to view the full-text of this paper.

Comput. Chem. 1993 Jun;17(2):149-163. Statistics of local complexity in amino acid sequences and sequence databases. Wootton JC, Federhen S. Full-text at ScienceDirect

Note: this paper is supplementary and thus OPTIONAL. You might need a subscription to the Computers & Chemistry magazine in order to view the full-text of this paper.

Recommended http://www.ncbi.nlm.nih.gov/pmc/articles/PMC290255/ (Recommended by Sebastien Beaune)

Readings

Required

Nature Biotechnology 22, 1315 - 1316 (2004). What is a hidden Markov model? Sean R Eddy. http://www.fing.edu.uy/~alopeza/biohpc/papers/hmm/Eddy-What-is-a-HMM-ATG4-preprint.pdf

Optional

Nucleic Acids Res. 1998 Jan 1;26(1):320-2. Pfam: multiple sequence alignments and HMM-profiles of protein domains. Sonnhammer EL, Eddy SR, Birney E, Bateman A, Durbin R. http://nar.oxfordjournals.org/content/26/1/320.full.pdf+html

## Markov Model

Andrei Andreyevich Markov (1856-1922). A Markov chaini describes a discrete stochastic process at successive times. The transitions from one state to any of all states, including itself, are governed by a probability distribution.
- $P(X_t | X_1...X_{t-1}) = P(X_i | X_{t-m}...X_{t-1})$
  - $X_t = f(X_{t-1}, X_{t-2}, ..., X_{t-m})$
- A chain of random variables in whick the next one depends (only) on the current one
  - $-P(X_t | X_{1}...X_{t-1}) = P(X_{i} | X_{i-1})$

Transition Probability
- $a_{kl} = P(X_{t} = S_{l} | X_{t-1} = S_{k})$, state k transite to state l.
- $a_{lk} = P(X_{t} = S_{k} | X_{t-1} = S_{l})$, state l transite to state k.
- $a_{kl}$ is not necassarily equal to $a_{lk}$

Hidden Markov Model (HMM)
- In addition to State Transition Probability, each state of HMM has a probability distribution over the possible output tokens (Emission Probability)
- Thus, a HMM is consist of two strings of information
  - The state path, which is not directly visible
  - The token path (emitted sqquence), based on the observable token path, we can infer the underling state path
- Transition and Emmision probability.
- Given a HMM, a sequence of tokens could be generated as following:
  - when we "visit" a state, we emit a token from the state's emission probability distribution
  - then, we choose which state to visit next, according to the state's transition probability distribution
    - transition probability, $a_{kl} = P(x_{t} = S_{l} | x_{t-1} = S_{k}$
    - emission probability, $e_{k}(b) = P(y_{i} = b | x_{i} = S_{k})$
  - $P(X, Y) = \prod_{i=1}^L(e_{x_{i}}(y_{i}*a_{x_{i}x_{i+1}})$
- M: Match (not necessarily identical); X: Insert at sequence X (delete at sequence Y); Y: Insert at sequence Y (delete at sequence X), $\delta$: Gap open; $\epsilon$: Gap extension.
  ||M|X|Y|
  |:------:|:------:|:------:|:------:|
  |**M**|$1-2\delta$|$\delta$|$\delta$|
  |**X**|$1-\epsilon$|$\epsilon$|0|
  |**Y**|$1-\epsilon$|0|0|

Sequence alignment with HMM
- Each "token" of the HMM is an aligned pair of two residues (M state), or of a residue and a gap (X or Y state)
  - transition and emission probabilityes define the probability of each aligned pair of sequences
- Based on the HMM, each alignment of two sequences can be assigned with a probability
  - given two input sequences, we look for an alignment with the maximum probability
    - arg max(P(S1, S2, ali))
    - $P(X, Y, ali) = max(P_{M}(n,m), P_{X}(n,m), P_{Y}(n,m))$

Given the nature of HMM, many different state paths can give rise to the same token sequence, so we can simply sum up them together to get the full probability of a given token sequence: $P(X, Y) = \sum_{ali}(X, Y, ali)$

HMM: as a predictor

The Most Simple Gene Predictor (MSGP): Given a stretch of genomic sequence, where are the coding regions and where are noncoding regions?

Training the model
- What we need to train?
  - transition probabilities between states
  - emission probabilities for each state
- Estimate probabilities from known "training set"
  - an annotated genomic region, with coding/noncoding sequences labeled
- $P_{coding}(i+1) = e_{coding}(x_{i+1}) max_{k \in (coding, noncoding)}(P_{k}(i)a_{k->coding})$
- $P_{noncoding}(i+1) = e_{noncoding}(x_{i+1}) max_{k \in (coding, noncoding)}(P_{k}(i)a_{k->noncoding})$
- $P(X, S) = max(P_{coding}(n), P_{noncoding}(n))$
- Logarithmic transformation: ease calculation: $Log(a*b) = Log(a) + Log(b)$
- "CGAAAAAATCG" will get the result "NNCCCCCCNNN" usign MSGP.

GenScan:
- Chris Burge (1996): A 27-state sem-HMM
- A simpler model: 19-state
- A model taking UTR introns into account: 35-state

By decoupling states and tokens, Hidden Markov Model (HMM) provides a sound probability framework to model complex biological sequences.

Readings Required

Nature Biotechnology 22, 1315 - 1316 (2004). What is a hidden Markov model? Sean R Eddy. http://www.fing.edu.uy/~alopeza/biohpc/papers/hmm/Eddy-What-is-a-HMM-ATG4-preprint.pdf

Optional

Nucleic Acids Res. 1998 Jan 1;26(1):320-2. Pfam: multiple sequence alignments and HMM-profiles of protein domains. Sonnhammer EL, Eddy SR, Birney E, Bateman A, Durbin R. http://nar.oxfordjournals.org/content/26/1/320.full.pdf+html

## Next Generation Sequencing (NGS)

Read: A short DNA fragment which is a readout by sequencer (in FASTQ format):
- DNA sequence (symbols)
- Quality information

Quality: Given p = the probability of a base calling is wrong, its Quality Score can be written as:
- Q = $-10*\log_{10}(p)$

Association Study: for the given phenotypic trait, "functional variants" could be identified by comparing allele frequencies at hundreds of thousands of polymorphic sites, i.e allele A is associated with phenotypic trait P if (and only if) people who have P also have A more (or less) often than would be predicted from individual frequencies of A and P in the assessed population.

RNA-Seq: Explore the transcriptome. "A transcriptome is a collection of all the transcripts present in a given cell." (NHGRI factsheet, NIH, US)

Chromatin ImmunoPrecipitation Sequencing (ChIP-Seq): Profile Protein-DNA interaction. (deep sequencing technology)

Reads Mapping and Variants Calling: the first step of deep sequencing

Mapping: Input data
- Reference Genome
  - Nucleotide
  - Length: humdreds of Mb per chromosome
  - ~3 Gb in total (for human genome)
- Reads
  - Nucleotide, with various qualities (relatively high error rate: 1e-2~1e-5)
  - Length: 36~80 bp per read
  - Hundreds of Gbs per run

 "Embeded" Alignment: One sequence is "embeded" in the other sequence (NGS Reads, PCR primer, etc.) What we need here is actually a hybrid "global-local" alignment.
 - "Global" for short sequence (i.e. NGS Read)
 - But "Local" fro long sequence (i.e. Reference Genome)
 - In particular, the surrounding "overhang" gaps should be not penalized

BLAST Ideas: Seeding-and-extending
1. Find matches (seed) between the query and subject 
2. Extend seed into High Scoring Segment Pairs (HSPs) 
   - Run Smith-Waterman algorithm on the specified region only
3. Assess the reliability of the alignment

The speed is a big mater. Using Prefix tree or Suffix tree. 

HBS: A native hash function

Pigeonhole principle

Mapping Quality: Given reference sequence z (length L), a read sequence x (length l), u is the alignment position of x on z, the probability that z actually coming from the position u is p(z|x,u)

$p(z|x,u) = \prod_{mismatch}p(z_{i})$

$SQ(U) = \log (p(z | x, u)) = \sum_{mismatch}p(z_{i}) = \sum_{mismatch}Q(z_{i})$

If we assume that a uniform NULL model, i.e. the read can randomly come from all possible positions with equal probability, then the error of mapping to a specified position u could be written as:

$E(u) = \frac{SQ(u)}{\sum_{i}SQ(i)}$

Genetic Variants
- SNV: single nucleotide variant
  - Substitution (SNP)
  - Indel: insertion/deletion
- Structural Variation (SV)
  - Large-scale insertion/deletion
  - Inversion
  - Translocation
  - Copy Number Variation (CNV)

SNP calling is not genotyping: source: Nature Reviews Genetics 12, 443-451.
- SNP calling aims to determine in which positions there are polymorphisms or in which positions at least one of the bases differs from a reference sequence
- Genotype calling is the process of determining the genotype for each individual and is typically only done for positions in which a SNP or a 'variant' has already been called

Counting: an intuitive (and naive) approach
- counting high-confident, non-reference allele (i.e. Quality >= 20)
  - Freq < 20% or > 80%: homozygous genotype
  - Otherwise: heterozygous
- Works well for "deeply sequenced regions" (DSR), i.e. depth > 25x
  - But suffer from under-calling of heterozygous genotypes for low-coverage regions
  - And can't give an objective measurement for reliability

  Source: Nature Reviews Genetics 12, 443-451.

  A simple probabilistic model for genotyping
  1. For a diploid genome, there will be at most two different alleles (A and a) observed at a given site:
    - 3 possible genotypes: <A,A>, <A,a>, <a,a>
    - number of A: k, number of a: n-k
2. Then, the probability for each genotypes is
  - P(D | <A,A>) = the probability that we have (n-k) sequencing errors at this site $\prod_{n-k}P(x_{i})$
  - Similarly, we can see the P(D | <a,a>) = $\prod_{k}P(x_{i})$
  - P(D | <A,a>) = 1- (P(D | <A,A>) + P(D | <a,a>))
3. Bayes Formula can be further employed to calculate posterior probabilities, i.e. P(<A,A> | D), P(<a,a> | D), P(<A,a> | D) if we can estimate the prior probabilities P(<A,A>), P(<a,a>) and P(<A,a>).

outline:
- BWA & BWT (Burrows-Wheeler transform) algorithm
  - BWT is the compression algorithm used in BWA 
  - Lossless compression
  - Sort and transform the char matrix with string rotation
  - Reverse-char method was utilized for match
  - Cannot handle gap
- Variant caller
  - samtools: mpileup + bcftools
  - GATK (Genome Analysis ToolKit): UnifiedGenotyper, HaplotypeCaller
  - [ref](http://www/broadinstitute.org/gatk/guide/best-practices)
- Demonstration

Introduction of Likelihood and Bayesian approach

Genotyper of MAQ and SNVMix

Readings Required

Nat Biotechnol. 2009 May;27(5):455-7. How to map billions of short reads onto genomes. Trapnell C, Salzberg SL. PMID: 19430453

Nat Rev Genet. 2010 Jan;11(1):31-46. Sequencing technologies - the next generation. Metzker ML. PMID: 19997069

An Introduction to Hidden Markov Models for Biological Sequences by Anders Krogh. Optional Nat Rev Genet. 2011 Jun;12(6):443-51. Genotype and SNP calling from next-generation sequencing data. Nielsen R, Paul JS, Albrechtsen A, Song YS. PMID: 21587300

Nat Rev Genet. 2013 Jul;14(7):460-70. Sequencing studies in human genetics: design and interpretation. Goldstein DB, Allen A, Keebler J, Margulies EH, Petrou S, Petrovski S, Sunyaev S. PMID: 23752795

## Variant Databases

Where did our genetic variations come from?
- inherited from parents
- de novo mutations
- somatic mutations

Types of genetic variations in a human genome
- Chromosomal aneuploidy
- Structural Variations (SVs): deletion, tandem duplication, interspersed duplication (Copy Number Variations (CNVs)), mobile-element insertion, novel sequence insertion, inversion, translocation.
- Indel - short Insertion/Deletion
  - Within intergenic/intronic regions
    - Non-frameshifting
    - Frameshifting
- SNV - Single nucleotide variation: there are about 3 million SNVs in one person's genome, equicalent to ~1/1000 frequency. promoter variant, coding variant, UTR variant, intronic variant, intergenic variants, non-coding RNA variant.
  - SNVs within coding regions:
    - Stop codon gain (nonsense)
    - Non-synonymous (nissense)
    - Synonymous (same sense/silent)
    - Affect splicing
    - Stop codon loss
  - Nonsense SNVs are usually considered deleterious
    - even though it is not always the case...
  - synonymous, intronic, and intergenic variations are often ignored
    - however, according to GWAS studies, 88% of trait-associated variants of weak effect are non-coding
    - they remain under-studied and new methods are needed
  - Most research so far had focused on missense SNVs
  - Known deleterious mutations are enriched in missense mutations
    - ~50 of all known mutations of Mendelian disorders are missense mutations. asscertainment bias?
  - Although missense SNVs change the protein sequence, many do not cause phenotypic changes.

Nomenclature
- Mutation: when the monor allele has a frequency less than 1% in the general population, we usually called it a mutation
- Polymorphism,: otherwise it is usually called a polymorphism. Sometimes the cut off of 5% is used.
- Variation/variant: mutations and polymorphisms together are called variations or variants.

Functional/Phenotypic "effects" of human genetic variations
- disease vs. normal
- deleterious vs. neutral
- personal trait differences (e.g., height)

- animal model phenotypic changes
- cellular phenotypic changes
 
 - protein function changes
 - protein structure changes

 Statistical and stochastic, not deterministic. Observations, not "truth"

Variant databases
- 1998 [dbSNP](http://www.ncbi.nlm.nih.gov/SNP/): build 138 contains genetic variations from 131 species
  - SNPs
  - Indels
  - multinucleotide polymorphism
  - microsatellite markers
  - short tandem repeats
  - heterozygous sequences
- 2010 dbVar
- 2012 [1000 genomes](http://www.1000genomes.org/)
- 1987 [OMIM](http://www.omim.org/): Online Mendelian Inheritance in Man
- 1996 [HGMD](http://www.hgmd.cf.ac.uk/): Human Gene Mutation Database, a comprehensive collection of gene mutations that underlie, or are associated with, human genetic diseases, manually curated from literature.
- 2004 COSMIC
- 2007 LSDBs
- 2012 ClinVar

Conservation-based and rule-based methods: SIFT & PolyPhen
- 1999, an early attempt based on BLOSUM substitution matrix
- Mor successful methods since 2001
  - Conservation-based methods (e.g., SIFT) [Sort Intolerant From Tolerant substitutions](http://sift.jcvi.org/)
    - Important positions (such as active sites) tend to be conserved in the protein family across species.
      - mutations at well-conserved positions tend to be deleterious
    - Some positions have a high degree of diversity across species
      - mutations at these positions tend to be neutral
  - Rule-based methods (e.g., PolyPhen)
  - Machine learning classifier-based methods (e.g., [PolyPhen2](http://genetics.bwh.harvard.edu/pph2/), [SAPRED](http://www.sapred.cbi.pku.edu.cn/): **S**ingle **A**mino aicd **P**olymorphisms disease-association **Pred**ictor)

Sequence search database: SWISS-PORT; BLAST, FASTA, PSI-blast, CLUSTAL; MOTIF; SIFT & PolyPhen.

Homology Modeling: [SWISS-MODEL](http://swissmodel.expasy.org/)

Wherever there are challenges, there are opportunities.

Readings required

Bioinformatics. 2007 Jun 15;23(12):1444-50. Finding new structural and sequence attributes to predict possible disease association of single amino acid polymorphism (SAP). Ye ZQ, Zhao SQ, Gao G, Liu XQ, Langlois RE, Lu H, Wei L. PMID: 17384424

Optional:

Annu Rev Biophys Biomol Struct. 2000;29:291-325. Comparative protein structure modeling of genes and genomes. Martí-Renom MA, Stuart AC, Fiser A, Sánchez R, Melo F, Sali A. PMID: 10940251

